be/src/exec/pipeline/sink/file_sink_operator.cpp

```diff
@@ -0,0 +1,300 @@
+};
+
+// this function will be executed in bthread
+int FileSinkBuffer::execute_io_task(void* meta, bthread::TaskIterator<const vectorized::ChunkPtr>& iter) {
+    FileSinkBuffer* file_sink_buffer = static_cast<FileSinkBuffer*>(meta);
+    for (; iter; ++iter) {
+        file_sink_buffer->process_chunk(iter);
+    }
+    return 0;
+}
+
+void FileSinkBuffer::process_chunk(bthread::TaskIterator<const vectorized::ChunkPtr>& iter) {
+    --_num_pending_chunks;
+    // close is already done, just skip
+    if (_is_finished) {
+        return;
+    }
+
+    // cancelling has happened but close is not invoked
+    if (_is_cancelled && !_is_finished) {
+        close(_state);
+        return;
+    }
+
+    if (!_is_writer_opened) {
+        if (Status status = process_writer_op([this]() { return _writer->open(_state); }); !status.ok()) {
+            set_io_status(status);
+            close(_state);
+            return;
+        }
+        _is_writer_opened = true;
+    }
+    auto chunk = *iter;
+    if (chunk == nullptr) {
+        // this is the last chunk
+        close(_state);
+        return;
+    }
+    if (Status status = process_writer_op([this, chunk]() { return _writer->append_chunk(chunk.get()); });
+        !status.ok()) {
+        set_io_status(status);
+        close(_state);
+    }
+}
+
+Status FileSinkBuffer::process_writer_op(std::function<Status()> op) {
+    // for some writers(eg. hdfs) that cannot be executed in bthread, we need to put them in pthread.
+    if (_writer->can_run_in_bthread()) {
+        return op();
+    } else {
+        auto ret = call_hdfs_scan_function_in_pthread(op);
+        return ret->get_future().get();
+    }
+}
+
+Status FileSinkBuffer::prepare(RuntimeState* state, RuntimeProfile* parent_profile) {
+    bool expected = false;
+    if (!_is_prepared.compare_exchange_strong(expected, true)) {
+        return Status::OK();
+    }
+
+    RETURN_IF_ERROR(state->exec_env()->result_mgr()->create_sender(state->fragment_instance_id(), 1024, &_sender));
+
+    _state = state;
+    _writer = std::make_shared<FileResultWriter>(_file_opts.get(), _output_expr_ctxs, parent_profile);
+    RETURN_IF_ERROR(_writer->init(state));
+
+    bthread::ExecutionQueueOptions options;
+    _exec_queue_id = std::make_unique<bthread::ExecutionQueueId<const vectorized::ChunkPtr>>();
+    int ret = bthread::execution_queue_start<const vectorized::ChunkPtr>(_exec_queue_id.get(), &options,
+                                                                         &FileSinkBuffer::execute_io_task, this);
+    if (ret != 0) {
+        return Status::InternalError("start execution queue error");
+    }
+    return Status::OK();
+}
+
+Status FileSinkBuffer::append_chunk(RuntimeState* state, const vectorized::ChunkPtr& chunk) {
+    Status status = get_io_status();
+    if (!status.ok()) {
+        return status;
+    }
+    if (bthread::execution_queue_execute(*_exec_queue_id, chunk) != 0) {
```

Comment:
what will happen if source operator execute faster than sink buffer

Reply:
> what will happen if source operator execute faster than sink buffer

chunks will be queued in the execution queue, if there are a log of queued chunks(>64), FileSinkOperator::need_input will return false to achieve the purpose of back pressure.